cli/src/main/java/org/aya/cli/library/LibraryModuleLoader.java

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2020-2021 Yinsen (Tesla) Zhang.
+// Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
+package org.aya.cli.library;
+
+import kala.collection.Seq;
+import kala.collection.SeqView;
+import kala.collection.immutable.ImmutableSeq;
+import org.aya.api.error.Reporter;
+import org.aya.api.error.SourceFileLocator;
+import org.aya.concrete.parse.AyaParsing;
+import org.aya.concrete.resolve.ResolveInfo;
+import org.aya.concrete.resolve.context.EmptyContext;
+import org.aya.concrete.resolve.module.FileModuleLoader;
+import org.aya.concrete.resolve.module.ModuleLoader;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+/**
+ * This module loader is used to load source/compiled modules in a library.
+ * For modules belonging to this library, both compiled/source modules are searched and loaded (compiled first).
+ * For modules belonging to dependencies, only compiled modules are searched and loaded.
+ * This is archived by not storing source dirs of dependencies to
+ * {@link LibraryModuleLoader#thisModulePath} and always trying to find compiled cores in
+ * {@link LibraryModuleLoader#thisOutRoot}.
+ *
+ * <p>
+ * Unlike {@link FileModuleLoader}, this loader only resolves the module rather than tycking it.
+ * Tyck decision is made by {@link LibraryCompiler} by judging whether the source file
+ * is modified since last build.
+ *
+ * @param thisModulePath Source dirs of this module, out dirs of all dependencies.
+ * @param thisOutRoot    Out dir of this module.
+ * @author kiva
+ * @see FileModuleLoader
+ */
+public record LibraryModuleLoader(
+  @NotNull Reporter reporter,
+  @NotNull SourceFileLocator locator,
+  @NotNull SeqView<Path> thisModulePath,
+  @NotNull Path thisOutRoot
+) implements ModuleLoader {
+  public static @NotNull Path resolveCompiledCore(@NotNull Path basePath, @NotNull Seq<@NotNull String> moduleName) {
+    var withoutExt = moduleName.foldLeft(basePath, Path::resolve);
+    return withoutExt.resolveSibling(withoutExt.getFileName() + ".ayac");
```

Comment:
![image](https://user-images.githubusercontent.com/16398479/142772527-62598771-4092-4c4f-8634-227f0103e004.png)

Can we merge these `+ "c"` things into one?